packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_prop/component_ref_prop.rs

```diff
@@ -0,0 +1,105 @@
+    /// Creates a ComponentRefs prop that returns the last child with component_type
+    pub fn new_from_last_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            component_to_select: Some(ComponentToSelect::Last),
+        }
+    }
+
+    /// Creates a ComponentRefs prop that returns the first child with component_type
+    pub fn new_from_first_matching_child(component_type: &'static str) -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::ComponentRefs {
+                container: PropSource::Me,
+                filter: Rc::new(ContentFilter::IsType(component_type)),
+            },
+            component_to_select: Some(ComponentToSelect::First),
+        }
+    }
+
+    pub fn new_self_ref() -> Self {
+        ComponentRefProp {
+            data_query: DataQuery::SelfRef,
+            component_to_select: None,
+        }
+    }
+}
+
+/// Structure to hold data generated from the data queries
+#[derive(TryFromDataQueryResults, Debug)]
+#[data_query(query_trait = DataQueries, pass_data = &ComponentRefProp)]
+struct RequiredData {
+    refs: PropView<prop_type::ContentRefs>,
+}
+impl DataQueries for RequiredData {
+    fn refs_query(arg: &ComponentRefProp) -> DataQuery {
+        arg.data_query.clone()
+    }
+}
+
+impl PropUpdater for ComponentRefProp {
+    type PropType = prop_type::ComponentRef;
+
+    fn data_queries(&self) -> Vec<DataQuery> {
+        vec![self.data_query.clone()]
+    }
+
+    fn calculate(&self, data: DataQueryResults) -> PropCalcResult<Self::PropType> {
+        // There are two options based on the data query that created us.
+        match self.data_query {
+            DataQuery::ComponentRefs { .. } => {
+                let required_data = RequiredData::try_from_data_query_results(data).unwrap();
+                let content_refs = required_data.refs;
+
+                if content_refs.value.is_empty() {
+                    return PropCalcResult::FromDefault(<Self as PropUpdater>::default(self));
+                }
+
+                let mut elements = content_refs.value.as_slice().iter().map(|content_ref|
+                    match content_ref {
+                        &ContentRef::Component(idx) => {
+                             ComponentRef::from(idx)
+                        },
+                        _ => unreachable!("data queries for element refs prop should return component graph nodes, found {:?}", content_ref.clone())
+                    }
+                );
+                let component = match self.component_to_select {
+                    Some(ComponentToSelect::First) => elements.next(),
+                    Some(ComponentToSelect::Last) => elements.last(),
+                    None => unreachable!("component_to_select should be Some"),
```

Comment:
With refusing `None`, the `new_self_ref()` will break. I guess we don't use it anywhere right now. Do we need `new_self_ref()`? If we remove `new_self_ref()`, then do we want to remove the option from `component_to_select`?

Reply:
Oh, I guess you are just missing a match arm for `self.data_query` to catch `DataQuery::SelfRef`.